File: app/classification.py
from __future__ import annotations

import re
from pathlib import Path


TOKEN_RE = re.compile(r"[a-z0-9]{2,}")


def clean_tokens(*values: object) -> list[str]:
    raw = " ".join(str(value or "").lower() for value in values)
    normalized = re.sub(r"[^a-z0-9]+", " ", raw)
    return TOKEN_RE.findall(normalized)


def _token_set(*values: object) -> set[str]:
    return set(clean_tokens(*values))


def _has_any(tokens: set[str], values: set[str]) -> bool:
    return bool(tokens & values)


def _single_match(tokens: set[str], mapping: list[tuple[str, set[str]]]) -> str | None:
    for label, aliases in mapping:
        if _has_any(tokens, aliases):
            return label
    return None


def _matches_all(tokens: set[str], *values: str) -> bool:
    return set(values).issubset(tokens)


def canonical_category_pair(category: str | None, subcategory: str | None = None) -> tuple[str | None, str | None]:
    main = " ".join(str(category or "").strip().split())[:80] or None
    sub = " ".join(str(subcategory or "").strip().split())[:80] or None
    if not main:
        return None, sub
    direct_map = {
        "Aria Blaze (Solo)": ("My Little Pony", "Aria Blaze"),
        "Sonata Dusk": ("My Little Pony", "Sonata Dusk"),
        "Dazzlings": ("My Little Pony", "Dazzlings"),
        "My Little Pony (Fluttershy)": ("My Little Pony", "Fluttershy"),
        "My Little Pony (Mane 6)": ("My Little Pony", "Mane 6"),
        "My Little Pony (Equestria Girls)": ("My Little Pony", "Equestria Girls"),
        "KPOP Demon Hunters (Huntrix)": ("KPOP Demon Hunters", "Huntrix"),
        "KPOP Demon Hunters (Mira)": ("KPOP Demon Hunters", "Mira"),
        "Resident Evil (Leon)": ("Resident Evil", "Leon"),
        "Final Fantasy (Cloud)": ("Final Fantasy", "Cloud Strife"),
    }
    return direct_map.get(main, (main, sub))


def infer_category_pair(
    *,
    filename: str,
    media_kind: str,
    title: str | None = None,
    current_category: str | None = None,
    size: tuple[int, int] | None = None,
) -> tuple[str, str | None]:
    stem = Path(filename).stem.lower()
    current_main, current_sub = canonical_category_pair(current_category)
    tokens = _token_set(stem, title, current_main, current_sub)

    mlp_tokens = {
        "mlp", "pony", "ponies", "equestria", "rainbooms", "mane", "cutie", "cmc",
        "twilight", "pinkie", "fluttershy", "rarity", "applejack", "scootaloo",
        "rainbow", "dash", "derpy", "sunset", "starlight", "trixie", "celestia",
        "luna", "discord", "spike",
    }
    dazzlings_tokens = {"dazzlings", "adagio", "aria", "sonata"}
    fnaf_tokens = {
        "fnaf", "freddy", "bonnie", "chica", "foxy", "roxanne", "roxy", "fazbear",
        "animatronic", "frenni", "bonfie",
    }
    neptunia_tokens = {"neptunia", "neptune", "nepgear", "noire", "blanc", "vert", "uzume", "plutia"}
    xenoblade_tokens = {"xenoblade", "pyra", "mythra", "nia", "mio", "eunie", "taion", "lanz", "noah"}
    sonic_tokens = {"sonic", "tails", "amy", "shadow", "knuckles", "rouge"}
    resident_evil_tokens = {"resident", "evil", "leon", "ada", "jill", "claire", "wesker", "ashley"}
    final_fantasy_tokens = {"final", "fantasy", "cloud", "strife", "tifa", "aerith", "sephiroth", "ff7", "ffvii"}
    kpop_demon_hunters_tokens = {"kpop", "demon", "hunters", "huntrix", "mira", "zoey", "rumi"}
    cartoon_tokens = {"boondocks", "cartoon", "anime"}
    meme_tokens = {"meme", "memes", "funny", "reaction"}

    mlp_subcategories = [
        ("Aria Blaze", {"aria"}),
        ("Sonata Dusk", {"sonata"}),
        ("Adagio Dazzle", {"adagio"}),
        ("Fluttershy", {"fluttershy", "flutter", "fluttertone", "flutterwitch", "flutterfrolic", "appleshy"}),
        ("Twilight Sparkle", {"twilight", "twilights"}),
        ("Rainbow Dash", {"dashie", "rainbowdash"}),
        ("Pinkie Pie", {"pinkie"}),
        ("Rarity", {"rarity"}),
        ("Applejack", {"applejack"}),
        ("Sunset Shimmer", {"sunset"}),
        ("Starlight Glimmer", {"starlight"}),
        ("Scootaloo", {"scootaloo"}),
        ("Derpy", {"derpy"}),
        ("Trixie", {"trixie"}),
        ("Princess Celestia", {"celestia"}),
        ("Princess Luna", {"luna"}),
        ("Discord", {"discord"}),
        ("Spike", {"spike"}),
    ]
    neptunia_subcategories = [
        ("Neptune", {"neptune"}),
        ("Nepgear", {"nepgear"}),
        ("Noire", {"noire"}),
        ("Blanc", {"blanc"}),
        ("Vert", {"vert"}),
        ("Uzume", {"uzume"}),
        ("Plutia", {"plutia"}),
    ]
    xenoblade_subcategories = [
        ("Pyra", {"pyra"}),
        ("Mythra", {"mythra"}),
        ("Nia", {"nia"}),
        ("Mio", {"mio"}),
        ("Eunie", {"eunie"}),
        ("Taion", {"taion"}),
        ("Lanz", {"lanz"}),
        ("Noah", {"noah"}),
    ]
    sonic_subcategories = [
        ("Sonic", {"sonic"}),
        ("Shadow", {"shadow"}),
        ("Amy", {"amy"}),
        ("Tails", {"tails"}),
        ("Knuckles", {"knuckles"}),
        ("Rouge", {"rouge"}),
    ]
    fnaf_subcategories = [
        ("Freddy", {"freddy"}),
        ("Bonnie", {"bonnie", "bonfie"}),
        ("Chica", {"chica"}),
        ("Foxy", {"foxy"}),
        ("Frenni", {"frenni"}),
        ("Roxanne Wolf", {"roxanne", "roxy"}),
    ]
    resident_evil_subcategories = [
        ("Leon", {"leon"}),
        ("Ada Wong", {"ada"}),
        ("Jill Valentine", {"jill"}),
        ("Claire Redfield", {"claire"}),
        ("Albert Wesker", {"wesker"}),
        ("Ashley Graham", {"ashley"}),
    ]
    final_fantasy_subcategories = [
        ("Cloud Strife", {"cloud", "strife"}),
        ("Tifa Lockhart", {"tifa"}),
        ("Aerith Gainsborough", {"aerith"}),
        ("Sephiroth", {"sephiroth"}),
    ]
    kpop_subcategories = [
        ("Huntrix", {"huntrix"}),
        ("Mira", {"mira"}),
        ("Zoey", {"zoey"}),
        ("Rumi", {"rumi"}),
    ]

    if Path(filename).suffix.lower() == ".gif":
        return "GIFs", None

    if _has_any(tokens, kpop_demon_hunters_tokens) or (current_main and current_main == "KPOP Demon Hunters"):
        return "KPOP Demon Hunters", _single_match(tokens, kpop_subcategories)

    if _has_any(tokens, resident_evil_tokens) or (current_main and current_main == "Resident Evil"):
        return "Resident Evil", _single_match(tokens, resident_evil_subcategories)

    if _has_any(tokens, final_fantasy_tokens) or (current_main and current_main == "Final Fantasy"):
        return "Final Fantasy", _single_match(tokens, final_fantasy_subcategories)

    if _has_any(tokens, neptunia_tokens) or (current_main and current_main == "Hyperdimension Neptunia"):
        neptunia_hits = [label for label, aliases in neptunia_subcategories if _has_any(tokens, aliases)]
        if len(neptunia_hits) > 1:
            return "Hyperdimension Neptunia", None
        return "Hyperdimension Neptunia", neptunia_hits[0] if neptunia_hits else None

    if _has_any(tokens, xenoblade_tokens) or (current_main and current_main == "Xenoblade"):
        xenoblade_hits = [label for label, aliases in xenoblade_subcategories if _has_any(tokens, aliases)]
        if len(xenoblade_hits) > 1:
            return "Xenoblade", None
        return "Xenoblade", xenoblade_hits[0] if xenoblade_hits else None

    if _has_any(tokens, fnaf_tokens) or (current_main and current_main == "FNAF"):
        fnaf_hits = [label for label, aliases in fnaf_subcategories if _has_any(tokens, aliases)]
        if len(fnaf_hits) > 1:
            return "FNAF", None
        return "FNAF", fnaf_hits[0] if fnaf_hits else None

    if (_has_any(tokens, sonic_tokens) and _has_any(tokens, mlp_tokens)) or (current_main and current_main == "Crossovers"):
        return "Crossovers", _single_match(tokens, sonic_subcategories) or _single_match(tokens, mlp_subcategories)

    if _has_any(tokens, mlp_tokens) or _has_any(tokens, dazzlings_tokens) or current_main == "My Little Pony":
        dazzling_hits: list[str] = []
        if "aria" in tokens:
            dazzling_hits.append("Aria Blaze")
        if "sonata" in tokens:
            dazzling_hits.append("Sonata Dusk")
        if "adagio" in tokens:
            dazzling_hits.append("Adagio Dazzle")
        if _has_any(tokens, dazzlings_tokens):
            if len(dazzling_hits) > 1 or "dazzlings" in tokens:
                return "My Little Pony", "Dazzlings"
            if dazzling_hits:
                return "My Little Pony", dazzling_hits[0]

        if _matches_all(tokens, "equestria", "girls") or _matches_all(tokens, "equestrian", "girls") or "eqg" in tokens or _matches_all(tokens, "rainbow", "rocks") or _matches_all(tokens, "crystal", "prep"):
            if dazzling_hits:
                return "My Little Pony", "Dazzlings"
            if _matches_all(tokens, "mane", "six") or _matches_all(tokens, "mane", "6"):
                return "My Little Pony", "Equestria Girls"
            return "My Little Pony", "Equestria Girls"

        mlp_character_hits: list[str] = []
        if _matches_all(tokens, "rainbow", "dash") or "dashie" in tokens or "rainbowdash" in tokens or "dash" in tokens:
            if not _matches_all(tokens, "rainbow", "rocks"):
                mlp_character_hits.append("Rainbow Dash")
        for label, aliases in mlp_subcategories:
            if label == "Rainbow Dash":
                continue
            if _has_any(tokens, aliases):
                mlp_character_hits.append(label)
        if _matches_all(tokens, "cutie", "mark", "crusaders") or "crusaders" in tokens or "cmc" in tokens:
            return "My Little Pony", "Cutie Mark Crusaders"
        if _matches_all(tokens, "mane", "6") or _matches_all(tokens, "mane", "six"):
            return "My Little Pony", "Mane 6"
        if len({hit for hit in mlp_character_hits if hit in {"Fluttershy", "Twilight Sparkle", "Rainbow Dash", "Pinkie Pie", "Rarity", "Applejack"}}) >= 2:
            return "My Little Pony", "Mane 6"
        if mlp_character_hits:
            unique_hits: list[str] = []
            for hit in mlp_character_hits:
                if hit not in unique_hits:
                    unique_hits.append(hit)
            if len(unique_hits) == 1:
                return "My Little Pony", unique_hits[0]
            if len(unique_hits) > 1:
                return "My Little Pony", None
        return "My Little Pony", None

    if _has_any(tokens, sonic_tokens) or (current_main and current_main == "Sonic"):
        return "Sonic", _single_match(tokens, sonic_subcategories)

    if _has_any(tokens, cartoon_tokens) or (current_main and current_main == "Cartoon"):
        return "Cartoon", None

    if _has_any(tokens, meme_tokens) or (current_main and current_main == "Memes"):
        return "Memes", None

    if size:
        width, height = size
        if width and height:
            ratio = width / max(1, height)
            if 0.9 <= ratio <= 1.1 and max(width, height) <= 2048:
                return "Profile Pictures", None
            if ratio < 0.85:
                return "Phone Backgrounds", None
            if ratio >= 1.2:
                return "Desktop Backgrounds", None

    if media_kind == "video":
        return current_main or "Videos", current_sub

    if current_main:
        return current_main, current_sub

    return "Wallpapers", None

File: lua/src/classification.lua
-- Filename/title keyword-based category+subcategory inference, ported
-- faithfully from app/classification.py (pure string/set logic, no network
-- calls, no image decoding -- the one piece of the AI pipeline that's fully
-- self-contained and cheap to port in full). Used standalone as the
-- heuristic fallback whenever the real AI vision pipeline
-- (app/ai_metadata.py, NOT yet ported -- see the Lua rewrite's final report)
-- is disabled, unavailable, or low-confidence.
local M = {}

local function lower_words(...)
  local parts = {}
  for _, v in ipairs({ ... }) do
    if v ~= nil and v ~= "" then parts[#parts + 1] = tostring(v):lower() end
  end
  return table.concat(parts, " ")
end

-- Mirrors clean_tokens(): lowercase, non-alnum runs -> single space, then
-- extract tokens of 2+ [a-z0-9] chars.
function M.clean_tokens(...)
  local raw = lower_words(...)
  local normalized = raw:gsub("[^%a%d]+", " ")
  local tokens = {}
  for tok in normalized:gmatch("%a%a+") do tokens[#tokens + 1] = tok end
  for tok in normalized:gmatch("%d%d+") do tokens[#tokens + 1] = tok end
  -- Python's regex `[a-z0-9]{2,}` allows mixed alnum runs (e.g. "ff7"); the
  -- two gmatch passes above miss those, so also scan generically.
  tokens = {}
  for tok in normalized:gmatch("[%a%d][%a%d]+") do tokens[#tokens + 1] = tok end
  return tokens
end

local function token_set(...)
  local set = {}
  for _, tok in ipairs(M.clean_tokens(...)) do set[tok] = true end
  return set
end

local function has_any(tokens, values)
  for v, _ in pairs(values) do
    if tokens[v] then return true end
  end
  return false
end

local function to_set(list)
  local s = {}
  for _, v in ipairs(list) do s[v] = true end
  return s
end

local function single_match(tokens, mapping)
  for _, entry in ipairs(mapping) do
    if has_any(tokens, entry[2]) then return entry[1] end
  end
  return nil
end

local function matches_all(tokens, ...)
  for _, v in ipairs({ ... }) do
    if not tokens[v] then return false end
  end
  return true
end

local DIRECT_MAP = {
  ["Aria Blaze (Solo)"] = { "My Little Pony", "Aria Blaze" },
  ["Sonata Dusk"] = { "My Little Pony", "Sonata Dusk" },
  ["Dazzlings"] = { "My Little Pony", "Dazzlings" },
  ["My Little Pony (Fluttershy)"] = { "My Little Pony", "Fluttershy" },
  ["My Little Pony (Mane 6)"] = { "My Little Pony", "Mane 6" },
  ["My Little Pony (Equestria Girls)"] = { "My Little Pony", "Equestria Girls" },
  ["KPOP Demon Hunters (Huntrix)"] = { "KPOP Demon Hunters", "Huntrix" },
  ["KPOP Demon Hunters (Mira)"] = { "KPOP Demon Hunters", "Mira" },
  ["Resident Evil (Leon)"] = { "Resident Evil", "Leon" },
  ["Final Fantasy (Cloud)"] = { "Final Fantasy", "Cloud Strife" },
}

local function trim_collapse(s)
  s = tostring(s or "")
  s = s:gsub("^%s+", ""):gsub("%s+$", ""):gsub("%s+", " ")
  return s
end

-- Mirrors canonical_category_pair().
function M.canonical_category_pair(category, subcategory)
  local main = trim_collapse(category):sub(1, 80)
  local sub = trim_collapse(subcategory):sub(1, 80)
  if main == "" then main = nil end
  if sub == "" then sub = nil end
  if not main then return nil, sub end
  local mapped = DIRECT_MAP[main]
  if mapped then return mapped[1], mapped[2] end
  return main, sub
end

local function stem_of(filename)
  filename = tostring(filename or "")
  local base = filename:match("([^/\\]+)$") or filename
  local stem = base:match("^(.*)%.[^.]+$") or base
  return stem:lower()
end

local function ext_of(filename)
  filename = tostring(filename or "")
  local base = filename:match("([^/\\]+)$") or filename
  return (base:match("(%.[^.]+)$") or ""):lower()
end

-- Mirrors infer_category_pair(). Returns (category_name, subcategory_name|nil).
function M.infer_category_pair(opts)
  local filename = opts.filename or ""
  local media_kind = opts.media_kind or ""
  local title = opts.title
  local current_category = opts.current_category
  local size = opts.size -- {width, height} or nil

  local current_main, current_sub = M.canonical_category_pair(current_category)
  local tokens = token_set(stem_of(filename), title, current_main, current_sub)

  local mlp_tokens = to_set({
    "mlp", "pony", "ponies", "equestria", "rainbooms", "mane", "cutie", "cmc",
    "twilight", "pinkie", "fluttershy", "rarity", "applejack", "scootaloo",
    "rainbow", "dash", "derpy", "sunset", "starlight", "trixie", "celestia",
    "luna", "discord", "spike",
  })
  local dazzlings_tokens = to_set({ "dazzlings", "adagio", "aria", "sonata" })
  local fnaf_tokens = to_set({
    "fnaf", "freddy", "bonnie", "chica", "foxy", "roxanne", "roxy", "fazbear",
    "animatronic", "frenni", "bonfie",
  })
  local neptunia_tokens = to_set({ "neptunia", "neptune", "nepgear", "noire", "blanc", "vert", "uzume", "plutia" })
  local xenoblade_tokens = to_set({ "xenoblade", "pyra", "mythra", "nia", "mio", "eunie", "taion", "lanz", "noah" })
  local sonic_tokens = to_set({ "sonic", "tails", "amy", "shadow", "knuckles", "rouge" })
  local resident_evil_tokens = to_set({ "resident", "evil", "leon", "ada", "jill", "claire", "wesker", "ashley" })
  local final_fantasy_tokens = to_set({ "final", "fantasy", "cloud", "strife", "tifa", "aerith", "sephiroth", "ff7", "ffvii" })
  local kpop_demon_hunters_tokens = to_set({ "kpop", "demon", "hunters", "huntrix", "mira", "zoey", "rumi" })
  local cartoon_tokens = to_set({ "boondocks", "cartoon", "anime" })
  local meme_tokens = to_set({ "meme", "memes", "funny", "reaction" })

  local mlp_subcategories = {
    { "Aria Blaze", to_set({ "aria" }) },
    { "Sonata Dusk", to_set({ "sonata" }) },
    { "Adagio Dazzle", to_set({ "adagio" }) },
    { "Fluttershy", to_set({ "fluttershy", "flutter", "fluttertone", "flutterwitch", "flutterfrolic", "appleshy" }) },
    { "Twilight Sparkle", to_set({ "twilight", "twilights" }) },
    { "Rainbow Dash", to_set({ "dashie", "rainbowdash" }) },
    { "Pinkie Pie", to_set({ "pinkie" }) },
    { "Rarity", to_set({ "rarity" }) },
    { "Applejack", to_set({ "applejack" }) },
    { "Sunset Shimmer", to_set({ "sunset" }) },
    { "Starlight Glimmer", to_set({ "starlight" }) },
    { "Scootaloo", to_set({ "scootaloo" }) },
    { "Derpy", to_set({ "derpy" }) },
    { "Trixie", to_set({ "trixie" }) },
    { "Princess Celestia", to_set({ "celestia" }) },
    { "Princess Luna", to_set({ "luna" }) },
    { "Discord", to_set({ "discord" }) },
    { "Spike", to_set({ "spike" }) },
  }
  local neptunia_subcategories = {
    { "Neptune", to_set({ "neptune" }) }, { "Nepgear", to_set({ "nepgear" }) },
    { "Noire", to_set({ "noire" }) }, { "Blanc", to_set({ "blanc" }) },
    { "Vert", to_set({ "vert" }) }, { "Uzume", to_set({ "uzume" }) }, { "Plutia", to_set({ "plutia" }) },
  }
  local xenoblade_subcategories = {
    { "Pyra", to_set({ "pyra" }) }, { "Mythra", to_set({ "mythra" }) }, { "Nia", to_set({ "nia" }) },
    { "Mio", to_set({ "mio" }) }, { "Eunie", to_set({ "eunie" }) }, { "Taion", to_set({ "taion" }) },
    { "Lanz", to_set({ "lanz" }) }, { "Noah", to_set({ "noah" }) },
  }
  local sonic_subcategories = {
    { "Sonic", to_set({ "sonic" }) }, { "Shadow", to_set({ "shadow" }) }, { "Amy", to_set({ "amy" }) },
    { "Tails", to_set({ "tails" }) }, { "Knuckles", to_set({ "knuckles" }) }, { "Rouge", to_set({ "rouge" }) },
  }
  local resident_evil_subcategories = {
    { "Leon", to_set({ "leon" }) }, { "Ada Wong", to_set({ "ada" }) }, { "Jill Valentine", to_set({ "jill" }) },
    { "Claire Redfield", to_set({ "claire" }) }, { "Albert Wesker", to_set({ "wesker" }) }, { "Ashley Graham", to_set({ "ashley" }) },
  }
  local final_fantasy_subcategories = {
    { "Cloud Strife", to_set({ "cloud", "strife" }) }, { "Tifa Lockhart", to_set({ "tifa" }) },
    { "Aerith Gainsborough", to_set({ "aerith" }) }, { "Sephiroth", to_set({ "sephiroth" }) },
  }
  local kpop_subcategories = {
    { "Huntrix", to_set({ "huntrix" }) }, { "Mira", to_set({ "mira" }) },
    { "Zoey", to_set({ "zoey" }) }, { "Rumi", to_set({ "rumi" }) },
  }

  if ext_of(filename) == ".gif" then return "GIFs", nil end

  if has_any(tokens, kpop_demon_hunters_tokens) or current_main == "KPOP Demon Hunters" then
    return "KPOP Demon Hunters", single_match(tokens, kpop_subcategories)
  end
  if has_any(tokens, resident_evil_tokens) or current_main == "Resident Evil" then
    return "Resident Evil", single_match(tokens, resident_evil_subcategories)
  end
  if has_any(tokens, final_fantasy_tokens) or current_main == "Final Fantasy" then
    return "Final Fantasy", single_match(tokens, final_fantasy_subcategories)
  end
  if has_any(tokens, neptunia_tokens) or current_main == "Hyperdimension Neptunia" then
    local hits = {}
    for _, e in ipairs(neptunia_subcategories) do if has_any(tokens, e[2]) then hits[#hits + 1] = e[1] end end
    if #hits > 1 then return "Hyperdimension Neptunia", nil end
    return "Hyperdimension Neptunia", hits[1]
  end
  if has_any(tokens, xenoblade_tokens) or current_main == "Xenoblade" then
    local hits = {}
    for _, e in ipairs(xenoblade_subcategories) do if has_any(tokens, e[2]) then hits[#hits + 1] = e[1] end end
    if #hits > 1 then return "Xenoblade", nil end
    return "Xenoblade", hits[1]
  end
  if has_any(tokens, fnaf_tokens) or current_main == "FNAF" then
    local hits = {}
    for _, e in ipairs({
      { "Freddy", to_set({ "freddy" }) }, { "Bonnie", to_set({ "bonnie", "bonfie" }) },
      { "Chica", to_set({ "chica" }) }, { "Foxy", to_set({ "foxy" }) },
      { "Frenni", to_set({ "frenni" }) }, { "Roxanne Wolf", to_set({ "roxanne", "roxy" }) },
    }) do if has_any(tokens, e[2]) then hits[#hits + 1] = e[1] end end
    if #hits > 1 then return "FNAF", nil end
    return "FNAF", hits[1]
  end
  if (has_any(tokens, sonic_tokens) and has_any(tokens, mlp_tokens)) or current_main == "Crossovers" then
    return "Crossovers", single_match(tokens, sonic_subcategories) or single_match(tokens, mlp_subcategories)
  end

  if has_any(tokens, mlp_tokens) or has_any(tokens, dazzlings_tokens) or current_main == "My Little Pony" then
    local dazzling_hits = {}
    if tokens["aria"] then dazzling_hits[#dazzling_hits + 1] = "Aria Blaze" end
    if tokens["sonata"] then dazzling_hits[#dazzling_hits + 1] = "Sonata Dusk" end
    if tokens["adagio"] then dazzling_hits[#dazzling_hits + 1] = "Adagio Dazzle" end
    if has_any(tokens, dazzlings_tokens) then
      if #dazzling_hits > 1 or tokens["dazzlings"] then return "My Little Pony", "Dazzlings" end
      if #dazzling_hits > 0 then return "My Little Pony", dazzling_hits[1] end
    end

    if matches_all(tokens, "equestria", "girls") or matches_all(tokens, "equestrian", "girls")
      or tokens["eqg"] or matches_all(tokens, "rainbow", "rocks") or matches_all(tokens, "crystal", "prep") then
      if #dazzling_hits > 0 then return "My Little Pony", "Dazzlings" end
      return "My Little Pony", "Equestria Girls"
    end

    local mlp_character_hits = {}
    if (matches_all(tokens, "rainbow", "dash") or tokens["dashie"] or tokens["rainbowdash"] or tokens["dash"])
      and not matches_all(tokens, "rainbow", "rocks") then
      mlp_character_hits[#mlp_character_hits + 1] = "Rainbow Dash"
    end
    for _, e in ipairs(mlp_subcategories) do
      if e[1] ~= "Rainbow Dash" and has_any(tokens, e[2]) then mlp_character_hits[#mlp_character_hits + 1] = e[1] end
    end
    if matches_all(tokens, "cutie", "mark", "crusaders") or tokens["crusaders"] or tokens["cmc"] then
      return "My Little Pony", "Cutie Mark Crusaders"
    end
    if matches_all(tokens, "mane", "6") or matches_all(tokens, "mane", "six") then
      return "My Little Pony", "Mane 6"
    end
    local core_hits = {}
    local core_set = to_set({ "Fluttershy", "Twilight Sparkle", "Rainbow Dash", "Pinkie Pie", "Rarity", "Applejack" })
    local seen_core = {}
    for _, h in ipairs(mlp_character_hits) do
      if core_set[h] and not seen_core[h] then seen_core[h] = true; core_hits[#core_hits + 1] = h end
    end
    if #core_hits >= 2 then return "My Little Pony", "Mane 6" end
    if #mlp_character_hits > 0 then
      local unique_hits, seen = {}, {}
      for _, h in ipairs(mlp_character_hits) do
        if not seen[h] then seen[h] = true; unique_hits[#unique_hits + 1] = h end
      end
      if #unique_hits == 1 then return "My Little Pony", unique_hits[1] end
      if #unique_hits > 1 then return "My Little Pony", nil end
    end
    return "My Little Pony", nil
  end

  if has_any(tokens, sonic_tokens) or current_main == "Sonic" then
    return "Sonic", single_match(tokens, sonic_subcategories)
  end
  if has_any(tokens, cartoon_tokens) or current_main == "Cartoon" then return "Cartoon", nil end
  if has_any(tokens, meme_tokens) or current_main == "Memes" then return "Memes", nil end

  if size and size[1] and size[2] and size[1] > 0 and size[2] > 0 then
    local ratio = size[1] / math.max(1, size[2])
    if ratio >= 0.9 and ratio <= 1.1 and math.max(size[1], size[2]) <= 2048 then
      return "Profile Pictures", nil
    elseif ratio < 0.85 then
      return "Phone Backgrounds", nil
    elseif ratio >= 1.2 then
      return "Desktop Backgrounds", nil
    end
  end

  if media_kind == "video" then return current_main or "Videos", current_sub end
  if current_main then return current_main, current_sub end
  return "Wallpapers", nil
end

return M
